### DISCUSSION TITLE
Surprisingly hard to override default ignores in config file

### DISCUSSION THREAD

**Discussion by comex** [BurntSushi/ripgrep#2538] [General] [answered]
(I'm using ripgrep 13.0.0 on macOS; I verified that the behavior hasn't changed on `master`.)

When using ripgrep, I want to ignore some paths by default, such as `build.ninja`.  But sometimes I want to override that behavior and search everything.

Currently, I do this by putting `--glob` options in my ripgrep config file, such as:
```
--glob=!build.ninja
```

Another way to accomplish this might be to put an `--ignore-file` option in the config file.  This does seem to have the downside that there's no way to make the path relative to the home directory (neither `~` nor `$HOME` works).  I can't easily use an absolute path because I share my config files across multiple machines where my home directory is at different absolute paths.

Anyway, both types of ignores *can* be disabled at the command line.  `--glob` ignores can be canceled out by using another `--glob` argument (`-g'*'`).  `--ignore-file` ignores can be disabled using `--no-ignore-files`.

The problem (to the extent there is a problem) is that disabling them is surprisingly hard.  Both of these arguments are relatively hard to type (`-g'*'` requires some keyboard acrobatics, and `--no-ignore-files` is quite long).  Also, as a secondary issue, there is no single argument that does both.

Most of the other long "don't ignore this" arguments can be abbreviated with some number of `-u`s.  But there is no number of `-u`s that implies `-g'*'` or `--no-ignore-files`.  This is somewhat surprising given that the man page states twice that `rg -uuu` is like `grep -r`:

> Tip: to disable all smart filtering and make ripgrep behave a bit more like classical grep, use rg -uuu.

> rg -uuu is roughly equivalent to grep -r.

As for solutions: There are many possible design approaches, including deciding that this isn't a bug and doing nothing.  But for discussion's sake, here is what I'd do:

Add variations of the `--glob` and/or `--ignore-file` arguments, e.g. `--default-glob` and `--default-ignore-file`, that behave the same as the originals except that they're disabled by plain `--no-ignore`/`-u`.  For simplicity it might be best to have only `--default-ignore-file` but then add some way to make it `$HOME`-relative.

**Comment by BurntSushi** [comment, marked as answer]
>  This is somewhat surprising given that the man page states twice that `rg -uuu` is like `grep -r`:

It is. The problem is that you aren't using `rg -uuu`. You're using `rg -g '!build.ninja' -uuu`. None of the `-u` flags override explicit globs. Explicit globs always have priority. I imagine, for example, that some folks might put `-uuu` in their ripgrep config file and use, e.g., `-g '*.toml'` to achieve something similar to grep's` --include`.

But that's for `-g/--glob`... What about `--no-ignore-files`? Ah, the man page says:

```
This does not imply --no-ignore-files, since --ignore-file is specified explicitly as a command line argument.
```

> Add variations of the --glob and/or --ignore-file arguments, e.g. --default-glob and --default-ignore-file, that behave the same as the originals except that they're disabled by plain --no-ignore/-u. For simplicity it might be best to have only --default-ignore-file but then add some way to make it $HOME-relative.

I just can't really stomach these additions personally. The problem is that ripgrep's file filtering logic has a ridiculous number of options already. And adding stuff like this is just going to make it even harder to reason about. Short of re-thinking ripgrep's file filtering logic, it feels to me like it's roughly at its carrying capacity. Additions are possible, but they need to be really well justified. I don't think this one is enough of a problem to justify more flags.

I think there are other work-arounds you could explore:

* Use aliases instead of config files (or aliases with config files) to give you an extra dimension of configurability. (This is probably what I would do I think.)
* If you can recast your problem as "I want to search some `build.ninja` files and not others," then you could put them in `.ignore` or `.rgignore` files. This does result in a bummer where you have to do this everywhere though.
* You could remove the ignore rule but add a file type. Then you can use `-Tninja` (or whatever) to opt into ignoring them.

I realize none of these things are probably ideal. You can find problems with all of them. (I myself can see problems with each of them.) But they might be good enough to improve the situation. Not sure.
